File: nhom12/src/main/java/nhom12/example/nhom12/model/Voucher.java
package nhom12.example.nhom12.model;

import java.time.LocalDateTime;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import nhom12.example.nhom12.model.enums.VoucherDiscountType;
import nhom12.example.nhom12.model.enums.VoucherType;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

@Document(collection = "vouchers")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
public class Voucher extends BaseDocument {

  @Indexed(unique = true)
  private String code;

  private String description;
  private VoucherType type;
  private VoucherDiscountType discountType;
  private double discountValue;
  private Double maxDiscountAmount;
  private double minOrderValue;
  private Integer usageLimit;
  private int usedCount;
  private LocalDateTime startAt;
  private LocalDateTime endAt;
  private boolean active;
}

File: frontend/src/types/voucher.ts
export type VoucherType = 'PRODUCT' | 'SHIPPING';
export type VoucherDiscountType = 'PERCENTAGE' | 'FIXED_AMOUNT';

export interface AppliedVoucher {
  voucherId: string;
  code: string;
  type: VoucherType;
  discountType: VoucherDiscountType;
  discountValue: number;
  discountAmount: number;
}

export interface Voucher {
  id: string;
  code: string;
  description?: string;
  type: VoucherType;
  discountType: VoucherDiscountType;
  discountValue: number;
  maxDiscountAmount?: number | null;
  minOrderValue: number;
  usageLimit?: number | null;
  usedCount: number;
  startAt?: string | null;
  endAt?: string | null;
  active: boolean;
  usable: boolean;
}

export interface VoucherValidation {
  subtotal: number;
  originalShippingFee: number;
  shippingFee: number;
  productDiscount: number;
  shippingDiscount: number;
  totalDiscount: number;
  total: number;
  productVoucher?: AppliedVoucher | null;
  shippingVoucher?: AppliedVoucher | null;
}

export interface VoucherFormPayload {
  code: string;
  description?: string;
  type: VoucherType;
  discountType: VoucherDiscountType;
  discountValue: number;
  maxDiscountAmount?: number | null;
  minOrderValue: number;
  usageLimit?: number | null;
  startAt?: string | null;
  endAt?: string | null;
  active: boolean;
}
